Automatically install server binaries

jrsurge opened the pull request:
## Purpose and motivation

<!-- Discuss why you made the PR and what it attempts to do -->

This PR builds on #102 (and #101) and calls the installer when initialising the class.

As a result, users only have to install the Quark and recompile the class library.
As a bonus, the binaries will automatically update with the quark.

This may not be the desired approach, the implications of this are quite large.

It also adds a way to enable/disable the auto installation.

## Implementation

<!-- Discuss how you made the PR and why it works that way -->

This PR adds `ScinServerInstaller.setup` to the class initialisation.
Additionally, adds a friendly message that ScinServerInstaller is the thing that's about to run (before, potentially, a bunch of download messages appear).

Adds `ScinServerInstaller.enableAutoInstall` and `ScinServerinstaller.disableAutoInstall` - and `ScinServerInstaller.autoInstallEnabled`.

The disabling of the auto install is accomplished by checking for the file `disable_auto_install` in the binary directory for the quark.

Also, adds `Scintillator` utility class for reporting quark-wide information - and refactors existing code to utilise it.

## Types of changes

<!-- Delete as needed -->
* Enhancement
* Documentation
* Behaviour change

## Status

- [x] This PR is ready for review

jrsurge committed:
add friendly warning to installer

lnihlen commented:
I worry about unintended consequences, like we end up crash-looping somebody's setup in a pathological combination of some kind of bug in the install script and problems with class re-compilation or something.

So I was thinking of a few changes to the installer:

- [ ] Before doing anything the installer should check for the existence of a file in the `/bin` directory, maybe something like `/bin/disable_auto_install` or similar. If that file exists the installer should immediately exit normally.

- [ ] Please refactor the installer to use the `ScinServerOptions.quarkPath` function instead of rolling its own using the old Quark enumeration trick

- [ ] Consider making the installer "happy path" print nothing else to the log? Just to cut down on logspam?

jrsurge committed:
add auto installation of server binary

jrsurge committed:
refactor Scintillator info querying

- adds Scintillator class:
  responsible for version, path to quark, path to binary

jrsurge committed:
refactor installer to use Scintillator queries

jrsurge committed:
check for disable_auto_install file

jrsurge committed:
use binDir instead of binPath

it returns the directory for the binary, not the full path of the binary

jrsurge committed:
swap Scintillator.binPath for Scintillator.binDir

jrsurge committed:
move auto installation to installer

jrsurge committed:
print path for installer

jrsurge committed:
update help

jrsurge committed:
update tests to use Scintillator class
